### PR TITLE
Adds the Option to Concatenate Embeddings of Directed Graphs

### LINKED ISSUE #397: add option to concatenate directed embeddings before outputting

### PR DESCRIPTION
**Pull request by spencer-loggia**
#### Reference Issues/PRs

Fixes #397 

#### Explanation of changes
Adds option to concatenate directed embeddings before output. An optional parameter `concat` (default False) is added to  `baseEmbed.__init__`,  as well as the `__init__`  methods of it's children. When `baseEmbed.fit_transform_` is called and  concat is true,  returns the left and right latent matrices concatenated on axis 1. If `concat` is False it behaves the same as before, and returns a tuple of the left and right latent matrices. 

Also add tests of directed embedding to `test_spectral_embed.py` and `test_omni.py`, and refactors `test_mase` to make use of new concat parameter.

### CODE REVIEW

**Comment by bdpedigo** [comment]
@spencer-loggia will also need to write some tests for this, FYI

**Comment by spencer-loggia** [comment]
Ok, I'll add some tests, I hadn't originally because I didn't see any tests of embedding direct graphs to begin with

**Comment by spencer-loggia** [comment]
@bdpedigo it should be ready completely ready!

File: graspy/embed/base.py [hunk 1]
```diff
@@ -158,11 +165,10 @@ def fit_transform(self, graph, y=None):
 
         Returns
         -------
-        out : np.ndarray, shape (n_vertices, n_dimension) OR tuple (len 2)
-            Where both elements have shape (n_vertices, n_dimension)
-            A single np.ndarray represents the latent position of an undirected
-            graph, wheras a tuple represents the left and right latent positions
-            for a directed graph
+        out : np.ndarray OR length 2 tuple of np.ndarray.
+            if undirected then returns single np.ndarray of latent position, shape(n_vertices, n_dimension).
```
Comment (bdpedigo) [inline graspy/embed/base.py:169]: `n_components` not `n_dimensions`

File: graspy/embed/base.py [hunk 2]
```diff
@@ -158,11 +165,10 @@ def fit_transform(self, graph, y=None):
...
+            if directed, self.concat = True then concatenate latent matrices on axis 1, shape(n_vertices, 2*n_dimension).
```
Comment (bdpedigo) [inline graspy/embed/base.py:170]: can just say `concat` but want to wrap any variables in double backticks. look through some of the other documentation for examples of how to ref variables

File: graspy/embed/lse.py [hunk 3]
```diff
@@ -71,6 +71,9 @@ class LaplacianSpectralEmbed(BaseEmbed):
         node degree is added. If int or float, must be >= 0. Only used when
         ``form`` == 'R-DAD'.
 
+    concat : bool, optional (default False)
+        if graph is directed whether to concatenate each graph's embedding along axis 1.
```
Comment (bdpedigo) [inline graspy/embed/lse.py:75]: in this description and all others, describe that this is concatenation of the "out" and "in" latent positions for the directed graph

**Comment by bdpedigo** [comment]
@spencer-loggia looking good so far - a few more comments 
- will want to check the input argument, i.e. make sure that concat is a boolean. the code for this is kinda buried in selectSVD i think, at the moment, but can look at other code not in embed for examples of this kind of thing too
- do you know how to resolve merge conflicts? you have some on this branch that will need to be resolved. I recommend using `git merge`

**Review by bdpedigo — CHANGES_REQUESTED** [review]
see comments above

**Comment by spencer-loggia** [comment]
I can fix the conflicts, I'll rebase master into my branch to keep the history clean

**Comment by spencer-loggia** [comment]
@bdpedigo Codecov report is still failing, but says only +/- 0.02 %. Not really sure how to interpret this / whether it needs to be fixed, but anything new I've added is tested.

**Comment by alyakin314** [comment]
> @bdpedigo Codecov report is still failing, but says only +/- 0.02 %. Not really sure how to interpret this / whether it needs to be fixed, but anything new I've added is tested.

you can just create a test that triggers that warning using. similar to dis: https://github.com/neurodata/graspy/blob/23f5415f79e47582c92942de24ce6d5fcbeaca27/tests/test_spectral_embed.py#L106-L109 (but u prob want to do it in the test for base embed).

File: graspy/embed/omni.py [hunk 4]
```diff
@@ -215,9 +222,11 @@ def fit_transform(self, graphs, y=None):
 
         Returns
         -------
-        out : array-like, shape (n_graphs, n_vertices, n_components) if input
-            graphs were symmetric. If graphs were directed, returns tuple of
-            two arrays (same shape as above) where the first corresponds to the
-            left latent positions, and the right to the right latent positions
+        out : np.ndarray or length 2 tuple of np.ndarray.
+            if input graphs were symmetric, nndarray of shape (n_graphs, n_vertices, n_components).
```
Comment (bdpedigo) [inline graspy/embed/omni.py:226]: typo (nndarray)

File: graspy/embed/mase.py [hunk 5]
```diff
@@ -75,6 +75,11 @@ class MultipleASE(BaseEmbedMulti):
         a vector corresponding to the degree (or sum of edge weights for a
         weighted network) before embedding.
 
+    concat : bool, optional (default False)
+        if graph(s) are directed, whether to concatenate each graph's left and right (out and in) latent positions
```
Comment (bdpedigo) [inline graspy/embed/mase.py:79]: capitalize if

File: graspy/embed/lse.py [hunk 6]
```diff
@@ -71,6 +71,10 @@ class LaplacianSpectralEmbed(BaseEmbed):
         node degree is added. If int or float, must be >= 0. Only used when
         ``form`` == 'R-DAD'.
 
+    concat : bool, optional (default False)
+        if graph is directed, whether to concatenate left and right (out and in) latent positions along axis 1.
```
Comment (bdpedigo) [inline graspy/embed/lse.py:75]: also capitalize

File: graspy/embed/base.py [hunk 7]
```diff
@@ -158,11 +169,10 @@ def fit_transform(self, graph, y=None):
 
         Returns
         -------
-        out : np.ndarray, shape (n_vertices, n_dimension) OR tuple (len 2)
-            Where both elements have shape (n_vertices, n_dimension)
-            A single np.ndarray represents the latent position of an undirected
-            graph, wheras a tuple represents the left and right latent positions
-            for a directed graph
+        out : np.ndarray OR length 2 tuple of np.ndarray.
+            if undirected then returns single np.ndarray of latent position, shape(n_vertices, n_components).
```
Comment (bdpedigo) [inline graspy/embed/base.py:173]: also capitalize

**Review by bdpedigo — CHANGES_REQUESTED** [review]
just trigger a new build

**Comment by bdpedigo** [comment]
@spencer-loggia you'll need to trigger a new build (make some inconsequential commit) to proceed. Can always revert that commit after

**Comment by bdpedigo** [comment]
actually maybe what i just did is enough, we'll see

**Comment by spencer-loggia** [comment]
Seems good to go? Let me know if anything else is needed

**Comment by bdpedigo** [comment]
Sorry to nitpick but there are a bunch of places in the docs where you didn't capitalize first letter, could you just fix that?

**Review by bdpedigo — APPROVED** [review]
@spencer-loggia I am approving contingent on my one nitpicky doc change. 
@dwaynepryce I thought we should have this feature because it is a step towards the `topologic` behavior for embeddings. Can consider whether to make the default in the future

**Review by bdpedigo — APPROVED** [review]
